tinygrad/engine/schedule.py

```diff
@@ -43,15 +43,24 @@ class LBScheduleItem:
   var_vals: Dict[Variable, int] = field(default_factory=dict)
   metadata: List[Metadata] = field(default_factory=list)
 
+# *** Generic utils ***
+
+def lb_cache(func):
+  @functools.wraps(func)
+  def __wrapper(buf:LazyBuffer, st:ShapeTracker, *args, **kwargs):
+    if (buf, st) in (cache:=kwargs.get("cache", args[-1])): return cache[(buf, st)]
+    return cache.setdefault((buf, st), func(buf, st, *args, **kwargs))
+  return __wrapper
+
 # *** DAG transformation: List[LazyBuffer] -> ScheduleItem ***
 
+@lb_cache
 def _recursive_lazyop(buf:LazyBuffer, st:ShapeTracker, outputs:Tuple[LazyBuffer, ...], var_vals:Dict[Variable, int], inputs:Dict[LazyBuffer, int],
                       realizes:Dict[LazyBuffer, None], assign_targets:Dict[LazyBuffer, LazyBuffer],
                       reduce_info:Dict[Tuple[LazyBuffer, ShapeTracker], Tuple[ShapeTracker, Tuple[int, ...]]],
                       cache:Dict[Tuple[LazyBuffer, ShapeTracker], LazyOp]) -> LazyOp:
   """recursively create a lazyop"""
   if buf is not buf.base: st, buf = buf.st+st, buf.base
-  if (buf, st) in cache: return cache[(buf, st)]
```

Comment:
Does this work? The previous line modifies st

Reply:
yea.. that'd be a bug if you have a single view with two different st childs. We should have tests for this.